src/providers/google-vertex-ai/chatComplete.ts

```diff
@@ -0,0 +1,380 @@
+
+        if (shouldAppendEmptyModeChat) {
+          messages.push({ role: 'model', parts: [{ text: '' }] });
+        }
+
+        messages.push({ role, parts });
+        lastRole = role;
+      });
+
+      return messages;
+    },
+  },
+  temperature: {
+    param: 'generationConfig',
+    transform: (params: Params) => transformGenerationConfig(params),
+  },
+  top_p: {
+    param: 'generationConfig',
+    transform: (params: Params) => transformGenerationConfig(params),
+  },
+  top_k: {
+    param: 'generationConfig',
+    transform: (params: Params) => transformGenerationConfig(params),
+  },
+  max_tokens: {
+    param: 'generationConfig',
+    transform: (params: Params) => transformGenerationConfig(params),
+  },
+  stop: {
+    param: 'generationConfig',
+    transform: (params: Params) => transformGenerationConfig(params),
+  },
+  // https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/configure-safety-attributes
+  // Example payload to be included in the request that sets the safety settings:
+  //   "safety_settings": [
+  //     {
+  //         "category": "HARM_CATEGORY_DANGEROUS_CONTENT",
+  //         "threshold": "BLOCK_NONE"
+  //     },
+  //     {
+  //         "category": "HARM_CATEGORY_SEXUALLY_EXPLICIT",
+  //         "threshold": "BLOCK_ONLY_HIGH"
+  //     }
+  safety_settings: {
+    param: 'safety_settings',
+  },
+  tools: {
+    param: 'tools',
+    default: '',
+    transform: (params: Params) => {
+      const functionDeclarations: any = [];
+      params.tools?.forEach((tool) => {
+        if (tool.type === 'function') {
+          functionDeclarations.push(tool.function);
+        }
+      });
+      return [{ functionDeclarations }];
+    },
+  },
+};
+
+export interface GoogleErrorResponse {
+  error: {
+    code: number;
+    message: string;
+    status: string;
+    details: Array<Record<string, any>>;
+  };
+}
+
+interface GoogleGenerateFunctionCall {
+  name: string;
+  args: Record<string, any>;
+}
+
+interface GoogleGenerateContentResponse {
+  candidates: {
+    content: {
+      parts: {
+        text?: string;
+        functionCall?: GoogleGenerateFunctionCall;
+      }[];
+    };
+    finishReason: string;
+    index: 0;
+    safetyRatings: {
+      category: string;
+      probability: string;
+    }[];
+  }[];
+  promptFeedback: {
+    safetyRatings: {
+      category: string;
+      probability: string;
+      probabilityScore: number;
+      severity: string;
+      severityScore: number;
+    }[];
+  };
+  usageMetadata: {
+    promptTokenCount: number;
+    candidatesTokenCount: number;
+    totalTokenCount: number;
+  };
+}
+
+export const GoogleChatCompleteResponseTransform: (
+  response:
+    | GoogleGenerateContentResponse
+    | GoogleErrorResponse
+    | GoogleErrorResponse[],
+  responseStatus: number
+) => ChatCompletionResponse | ErrorResponse = (response, responseStatus) => {
+  // when error occurs on streaming request, the response is an array of errors.
+  if (
```

Comment:
With the new structural changes, now every provider can have a single error transform function which can be used in all other transformers. Please check some other providers to get an idea about this.